### ISSUE TITLE
Set X-Request-ID / client-request-id Header per-upload

### ISSUE THREAD

**Issue by michaeldeongreen** [Azure/azure-storage-js#54]
### Which service(blob, file, queue, table) does this issue concern?
blob
### Which version of the SDK was used?
10

### What problem was encountered?
Trying to figure out if you can set the X-Request-ID Header per blob upload.  We want to be able to set X-Request-ID/client request id Header for better logging and to sent to the EventGrid.

Based on the API, there is a BlockBlobUploadHeaders, but I am unsure 1) How to use it in my current code 2) If the requestId can be set 3) If this is even the X-Request-ID.

Here is my existing code but unsure if you can set the header here:

> const anonymousCredential = new AnonymousCredential();
    const pipeline = StorageURL.newPipeline(anonymousCredential);
    const serviceURL = new ServiceURL(
      {storageAssetInfo.uri}{storageAssetInfo.sharedAccessSignature},
      pipeline,
    );

> const containerName = storageAssetInfo.name;
    const containerURL = ContainerURL.fromServiceURL(serviceURL, containerName);
    const blobName = ${file.name}-${storageAssetInfo.name};
    const blobUrl = BlobURL.fromContainerURL(containerURL, blobName);
    const blockblobURL = BlockBlobURL.fromBlobURL(blobUrl);

> const options: IUploadToBlockBlobOptions = {
      blockSize: this.uploadToBlockBlobOptionsService.getBlockSize(file.size),
      parallelism: this.uploadToBlockBlobOptionsService.getParallelism(),
      maxSingleShotSize: this.uploadToBlockBlobOptionsService.getMaxSingleShotSize(file.size),
      progress: (transferProgressEvent: TransferProgressEvent) => {
        this.onFileUploadProgressChanged(transferProgressEvent, file, this.uploadProgressSource);
      },
    };

> const blobUploadCommonResponse = await uploadBrowserDataToBlockBlob(Aborter.none, file, blockblobURL, options);

### Steps to reproduce the issue?
None

### Have you found a mitigation/solution?
No

**Comment by XiaoningLiu** [comment]
@michaeldeongreen 

If you want to customize format of the client ids, like some prefixes or suffixes, you can do it with a new policy injected to the pipeline.

I created a wiki page demonstrate how to customize V10 storage SDK with pipeline and polices.
https://github.com/Azure/azure-storage-js/wiki/Customize-HTTP-Pipeline-and-Policies

If you want a customized `requestID` field in every request method options, it's not supported by design. 
I'd like to know is there any reason for you want to do this? But you always can create different BlockBlobURL objects with different pipelines assigned to achieve this.

**Comment by michaeldeongreen** [comment]
@XiaoningLiu , this is a feature requested by our client.


1) We will be using SDK 10 to upload to blob storage
2) Before the upload starts, the app will call an Web API to get metadata info ie SAS, container name etc tec
3) We would like a GUID to be apart of the metadata so we can set in the client-request-id
4) Once the upload completes, many events and workflows will follow (ie EventGrid, durable functions etc etc)
5) We would like to use the client-request-id as a means to track a unique id back to the original upload

I did some research over the weekend and saw:
BlobURL.setHTTPHeaders, but the client-request-id is not a header that can be set.

I also saw:
[](https://docs.microsoft.com/en-us/javascript/api/%40azure/storage-blob/blobsethttpheadersoptionalparams?view=azure-node-preview)
[](https://docs.microsoft.com/en-us/javascript/api/%40azure/storage-blob/blobsethttpheadersheaders?view=azure-node-preview)

But these objects are unavailable when you create a Blob via SAS Token.

I will look into a customized pipeline.
